independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/serialization/DynamicEntityWriter.java

```diff
@@ -41,22 +42,27 @@ public void write(ResteasyReactiveRequestContext context, Object entity) throws
             ServerHttpRequest vertxRequest = context.serverRequest();
             // first check and see if the resource method defined a media type and try to use it
             if ((context.getTarget() != null) && (context.getTarget().getProduces() != null)) {
+                ServerMediaType producesServerMediaType = context.getTarget().getProduces();
                 MediaType negotiatedMediaType = null;
                 List<String> accepts = context.getHttpHeaders().getRequestHeader(HttpHeaders.ACCEPT);
                 for (String accept : accepts) {
-                    negotiatedMediaType = context.getTarget().getProduces().negotiateProduces(accept).getKey();
+                    negotiatedMediaType = producesServerMediaType.negotiateProduces(accept).getKey();
                     if (negotiatedMediaType != null) {
                         break;
                     }
                 }
+                if (negotiatedMediaType == null) { // fallback to ensure that MessageBodyWriter is passed the proper media type
+                    negotiatedMediaType = producesServerMediaType
```

Comment:
How is this different from what is done in the loop above? Would it make sense to replace the loop with this call only?

Reply:
It uses the original HTTP headers, not the ones that filters can add to.

At some point this needs to be cleaned up, but not today :)